packages/@expo/cli/src/start/server/metro/metroVirtualModules.ts

```diff
@@ -0,0 +1,104 @@
+/**
+ * Copyright © 2024 650 Industries.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+import { FileSystem } from 'metro-file-map';
+import Bundler from 'metro/src/Bundler';
+
+type ExpoPatchedFileSystem = FileSystem & {
+  expoVirtualModules?: Map<string, Buffer>;
+};
+
+type ExpoPatchedBundler = Bundler & {
+  setVirtualModule: (id: string, contents: string) => void;
+  hasVirtualModule: (id: string) => boolean;
+};
+
+function ensureMetroBundlerPatchedWithSetVirtualModule(
+  bundler: Bundler & {
+    setVirtualModule?: (id: string, contents: string) => void;
+    hasVirtualModule?: (id: string) => boolean;
+  }
+): ExpoPatchedBundler {
+  if (!bundler.setVirtualModule) {
+    bundler.setVirtualModule = function (this: Bundler, id: string, contents: string) {
+      // @ts-expect-error: private property
+      const fs = ensureFileSystemPatched(this._depGraph._fileSystem);
+      fs.expoVirtualModules!.set(ensureStartsWithNullByte(id), Buffer.from(contents));
+    };
+    bundler.hasVirtualModule = function (this: Bundler, id: string) {
+      // @ts-expect-error: private property
+      const fs = ensureFileSystemPatched(this._depGraph._fileSystem);
+      return fs.expoVirtualModules!.has(ensureStartsWithNullByte(id));
+    };
+  }
+
+  return bundler as ExpoPatchedBundler;
+}
+
+function ensureStartsWithNullByte(id: string): string {
+  return id.startsWith('\0') ? id : `\0${id}`;
+}
+
+export function getMetroBundlerWithVirtualModules(bundler: Bundler): ExpoPatchedBundler {
+  // @ts-expect-error: private property
+  if (!bundler.transformFile.__patched) {
+    const originalTransformFile = bundler.transformFile.bind(bundler);
+
```

Comment:
I assume this is an expect-error for `__patched`. Have you considered adding this to the local types since we control the patched type anyway and can cast (since the additional property is also marked as optional)?

Could also use a symbol or a weak map market and see if that is a bit simpler.
Doesn't matter too much either